### PR TITLE
Github Workflows for CI

### PR DESCRIPTION
**Pull request by psieg**
Per suggestion, this sets up GH workflows to replace my jenkins instance, enabling forks to directly run them as well.

### CODE REVIEW

**Comment by psieg** [comment]
@zomfg ah, I just saw you already have some test setups as well in your fork. Feel free to expand these...

**Comment by zomfg** [comment]
oh
not much to expand, just few remarks from experience:
- I did not like `-latest` because it isn't, so I set versions manually for runners
- same for Qt on Windows
- on macOS they have multiple Xcodes, but it defaults to old stuff even on 11.0, so I added the bit that forces the latest installed Xcode
- dumping qmake/clang/gcc versions can be helpful
- `lscpu` or equivalent to keep track of SSE/AVX etc just in case
- [they already have reasonably recent versions of innosetup / openssl](https://github.com/actions/virtual-environments/blob/main/images/win/Windows2019-Readme.md#tools)
- actual artifact name isn't hard as you can see, but if you split everything it's not an issue I guess

**Comment by psieg** [comment]
Yeah, fixing versions in general is something I'm not sure about. One the one hand, more control, on the other, it needs to be manually kept up to date.

Ah. I saw some bug that suggested the preinstalled openssl had only binaries but not the libraries and headers etc. But since yours seems to be working without, I'll remove it.

**Comment by zomfg** [comment]
well if `-latest` was indeed latest or with minimal lag (for ex switch from `18.04` to `20.04` at `20.04.1` just like ubuntu itself does with auto upgrades), sure, or if they had `-previous`/`-super-stable` and `-current`/`-latest` or whatever...
but for a working Big Sur build you pretty much have to force every thing (since it was not working with older stuff #399 ) and if they wait as much as for ubuntu...

**Comment by zomfg** [comment]
also as you know 18.04 runs 5.9 which is EOL, so meh (that being said on desktop you can use the installer to switch to whatever, not sure if it works in server environment, but if that other action does it for windows it's probably doable elsewhere)
and msvc 15.something had some issues with some newer Qt syntax, so meh too (but I think it was fixed)

**Comment by zomfg** [comment]
`{{ matrix.os }}` wont work for you (not yet anyway), because I use it in generated ubuntu package names and you don't (and you don't have to, as is anyway)

**Comment by psieg** [comment]
Thanks. I was actually considering matrix for building both ubuntu versions.
Main concern for me is that users running Ubuntu LTS should be able to use the .deb from the releases page...

**Comment by zomfg** [comment]
I think docker is a better way to do multi distribution as it'll allow pretty much single matrix (as you can see in my `dpkg_build` job, which could be combined with `pacman_build` probably) to do everything, my scripts work ok it seems, but some cleaning is needed
they are both LTS so... sucks to be you :D
I did this https://github.com/zomfg/Lightpack/actions/runs/345185733 but I think it's not 100% inline with debian naming conventions so maybe few changes needed

**Comment by psieg** [comment]
hm, I'll stick with this for now, not really familiar with docker (nor debian). Feel free to add to it later

**Comment by zomfg** [comment]
example of release draft on tag

```yml
name: Build

on:
  push:
    # branches: [ master ]
    tags: [ '5.11.2.*' ] # release tag => create draft release

jobs:
  win_build:
    name: windows build
    runs-on: ubuntu-latest
    steps:
      - name: build
        run: |
          echo "this is some win build" > build-win.txt
          # zip --junk-paths build-win.zip build.txt

      - name: Upload Package
        uses: actions/upload-artifact@v2
        with:
          name: build-win.txt
          path: build-win.txt

  mac_build:
    name: mac build
    runs-on: ubuntu-latest
    steps:
      - name: build
        run: |
          echo "this is some mac build" > build-mac.txt
          # zip --junk-paths build-mac.zip build.txt

      - name: Upload Package
        uses: actions/upload-artifact@v2
        with:
          name: build-mac.txt
          path: build-mac.txt


  dpkg_build:
    name: ubuntu ${{ matrix.os }} build
    runs-on: ubuntu-latest
    strategy:
      matrix:
        os:
          - '18.04'
          - '20.04'
          - '20.10'

    steps:
      - name: build
        run: |
          echo "this is some ubuntu ${{ matrix.os }} build" > build-ubuntu-${{ matrix.os }}.txt
          # zip --junk-paths build-ubuntu-${{ matrix.os }}.zip build.txt

      - name: Upload Package
        uses: actions/upload-artifact@v2
        with:
          name: build-ubuntu-${{ matrix.os }}.txt
          path: build-ubuntu-${{ matrix.os }}.txt

  release:
    name: release on manual tag
    if: github.event_name == 'push'
    runs-on: ubuntu-20.04
    needs:
      - win_build
      - mac_build
      - dpkg_build

    steps:
      - name: download artifacts
        uses: actions/download-artifact@v2

      - name: Create Release
        # if: ${{ never() }}
        id: create_release
        uses: actions/create-release@v1
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          tag_name: ${{ github.ref }}
          release_name: Release ${{ github.ref }}
          # body: '${{ github.event.pull_request.title }} (#${{ github.event.pull_request.number }})'
          draft: true
          prerelease: false

      - name: Upload Release Asset
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          ASSET_URL: ${{ steps.create_release.outputs.upload_url }}
        run: |
          cat << EOF | python --
          import glob
          import mimetypes
          import os
          import requests

          for file in glob.iglob('**/*'):
            with open(file, 'rb') as asset:
              token=os.environ['GITHUB_TOKEN']
              headers = {
                'user-agent': os.environ['GITHUB_REPOSITORY'],
                'authorization': "token " + token,
                'content-type': mimetypes.guess_type(file)[0],
                'content-size': str(os.stat(file).st_size)
              }
              url=os.environ['ASSET_URL'].replace('{?name,label}','')
              requests.post(f"{url}?name={os.path.basename(file)}", headers=headers, data=asset)
          EOF

```

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+          name: firmware
+          path: Firmware/hex/*
+          
+  build-win:
+    name: Windows Build
+    runs-on: windows-2019
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          submodules: recursive
```
Comment (zomfg) [inline .github/workflows/ci.yml:53]: this thinks submodules are under the same user
`clone of 'https://github.com/zomfg/UpdateElevate.git' into submodule path 'D:/a/Lightpack/Lightpack/Software/UpdateElevate' failed`

**Pull request merged**